stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1385 @@
+    pub miner_signature: MessageSignature,
+    /// Recoverable ECDSA signature from the stacker set active during the tenure.
+    pub stacker_signature: MessageSignature,
+}
+
+#[derive(Debug, Clone)]
+pub struct NakamotoBlock {
+    pub header: NakamotoBlockHeader,
+    pub txs: Vec<StacksTransaction>,
+}
+
+pub struct NakamotoChainState;
+
+impl FromRow<NakamotoBlockHeader> for NakamotoBlockHeader {
+    fn from_row(row: &rusqlite::Row) -> Result<NakamotoBlockHeader, DBError> {
+        let version = row.get("version")?;
+        let chain_length_i64: i64 = row.get("chain_length")?;
+        let chain_length = chain_length_i64
+            .try_into()
+            .map_err(|_| DBError::ParseError)?;
+        let burn_spent_i64: i64 = row.get("burn_spent")?;
+        let burn_spent = burn_spent_i64.try_into().map_err(|_| DBError::ParseError)?;
+        let parent = row.get("parent")?;
+        let burn_view = row.get("burn_view")?;
+        let stacker_signature = row.get("stacker_signature")?;
+        let miner_signature = row.get("miner_signature")?;
+        let tx_merkle_root = row.get("tx_merkle_root")?;
+        let state_index_root = row.get("state_index_root")?;
+
+        Ok(NakamotoBlockHeader {
+            version,
+            chain_length,
+            burn_spent,
+            parent,
+            burn_view,
+            stacker_signature,
+            miner_signature,
+            tx_merkle_root,
+            state_index_root,
+        })
+    }
+}
+
+impl StacksMessageCodec for NakamotoBlockHeader {
+    fn consensus_serialize<W: std::io::Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &self.version)?;
+        write_next(fd, &self.chain_length)?;
+        write_next(fd, &self.burn_spent)?;
+        write_next(fd, &self.parent)?;
+        write_next(fd, &self.burn_view)?;
+        write_next(fd, &self.tx_merkle_root)?;
+        write_next(fd, &self.state_index_root)?;
+        write_next(fd, &self.miner_signature)?;
+        write_next(fd, &self.stacker_signature)?;
+
+        Ok(())
+    }
+
+    fn consensus_deserialize<R: std::io::Read>(fd: &mut R) -> Result<Self, CodecError> {
+        Ok(NakamotoBlockHeader {
+            version: read_next(fd)?,
+            chain_length: read_next(fd)?,
+            burn_spent: read_next(fd)?,
+            parent: read_next(fd)?,
+            burn_view: read_next(fd)?,
+            tx_merkle_root: read_next(fd)?,
+            state_index_root: read_next(fd)?,
+            miner_signature: read_next(fd)?,
+            stacker_signature: read_next(fd)?,
+        })
+    }
+}
+
+impl NakamotoBlockHeader {
+    pub fn block_hash(&self) -> BlockHeaderHash {
+        BlockHeaderHash::from_serializer(self)
+            .expect("BUG: failed to serialize block header hash struct")
+    }
+}
+
+impl NakamotoBlock {
+    /// Did the stacks tenure change on this nakamoto block? i.e., does this block
+    ///  include a TenureChange transaction?
```

Comment:
More specifically, does it include a `TenureChange` transaction for the current tenure?  A block can have multiple `TenureChange` transactions -- for example, it must include them for miners who never produced a block during their tenures, per the SIP.